### PR TITLE
feat: get top level dependecies for go lang

### PR DESCRIPTION
**Pull request by miiila**
- [ ] Ready for review
- [ ] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

### CODE REVIEW

File: lib/gomod-parser.ts [hunk 1]
```diff
@@ -173,3 +177,61 @@ export function toSnykVersion(v: ModuleVersion): string {
     return (v as ModuleExactVersion).exactVersion;
   }
 }
+
+export async function buildDepTreeFromImports(path: string = '.') {
+  const depTree: DepTree = {
+    name: 'tempName', // TODO: How do get name of the module?
```
Comment (kyegupov) [inline lib/gomod-parser.ts]: by using https://github.com/snyk/snyk-go-parser ?

Also, module or package?

File: lib/gomod-parser.ts [hunk 2]
```diff
@@ -173,3 +177,61 @@ export function toSnykVersion(v: ModuleVersion): string {
...
+    version: 'tempVersion', // TODO: How do get version of the module?
+    dependencies: {},
+  };
+  const goDepsOutput = await spawnGoListCommand(path);
+  if (goDepsOutput.includes('matched no packages')) {
+    return depTree;
+  }
+  const objectBorderRegexp = new RegExp(`}${EOL}{`, 'g');
+  const goDepsString = `[${goDepsOutput.replace(objectBorderRegexp, '},{')}]`;
+  const goDeps: [GoPackage] = JSON.parse(goDepsString);
```
Comment (kyegupov) [inline lib/gomod-parser.ts]: umm shouldn't the type be `GoPackage[]` ?

File: lib/gomod-parser.ts [hunk 3]
```diff
@@ -173,3 +177,61 @@ export function toSnykVersion(v: ModuleVersion): string {
...
+    depTree.dependencies![packageImport] = {
+      name: packageImport,
+      version: '<versionFromParentModule>', // TODO: Propagation of vesion from module is a separate task
+    };
+  }
+
+  return depTree;
+}
+
+function extractImports(goDeps: [GoPackage]): Set<string> {
```
Comment (kyegupov) [inline lib/gomod-parser.ts]: `GoPackage[]`?

File: lib/gomod-parser.ts [hunk 4]
```diff
@@ -173,3 +177,61 @@ export function toSnykVersion(v: ModuleVersion): string {
...
+  const goDepsImports: Set<string> = new Set();
```
Comment (kyegupov) [inline lib/gomod-parser.ts]: `const goDepsImports = new Set<string>();`

File: lib/types.ts [hunk 5]
```diff
@@ -64,3 +64,49 @@ export interface DepTree {
...
+  Standard?: boolean;     // is this package part of the standard Go library?
+  Stale?: boolean;     // would 'go install' do anything for this package?
+  StaleReason?: string;   // explanation for Stale==true
+  Root?: string;   // Go root or Go path dir containing this package
+  ConflictDir?: string;   // this directory shadows Dir in $GOPATH
+  BinaryOnly?: boolean;     // binary-only package: cannot be recompiled from sources
+  ForTest?: string;   // package is only for use in named test
+  Export?: string;   // file containing export data (when using -export)
+  Module?: GoModule;  // info about package's containing module, if any (can be nil)
+  Match?: [string]; // command-line patterns matching this package
```
Comment (kyegupov) [inline lib/types.ts]: `string[]` here and below

File: lib/types.ts [hunk 6]
```diff
@@ -64,3 +64,49 @@ export interface DepTree {
     [dep: string]: DepTree;
   };
 }
+
+export interface GoPackage {
```
Comment (kyegupov) [inline lib/types.ts]: a link to the original code/documentation would be nice

File: lib/gomod-parser.ts (hunk 1 above)
Comment (miiila) [reply lib/gomod-parser.ts]: Module in this case, this is the top level name, so user will be testing their module.

File: lib/gomod-parser.ts [hunk 7]
```diff
@@ -173,3 +177,61 @@ export function toSnykVersion(v: ModuleVersion): string {
...
+  const depTree: DepTree = {
+    name: 'tempName', // TODO: How do get name of the module?
+    version: 'tempVersion', // TODO: How do get version of the module?
+    dependencies: {},
+  };
+  const goDepsOutput = await spawnGoListCommand(path);
+  if (goDepsOutput.includes('matched no packages')) {
+    return depTree;
+  }
+  const objectBorderRegexp = new RegExp(`}${EOL}{`, 'g');
```
[the remark this answers is not in the document]
Comment (miiila) [reply lib/gomod-parser.ts]: It is a top level constant, but poorly described. It's https://nodejs.org/dist/latest-v8.x/docs/api/os.html#os_os_eol.

Is there any specific reason you prefer `\r?\n` over node api?

File: lib/gomod-parser.ts (hunk 2 above)
Comment (miiila) [reply lib/gomod-parser.ts]: It should, 🤦‍♂ on my side.

File: lib/types.ts (hunk 6 above)
Comment (miiila) [reply lib/types.ts]: Will try to find, this is taken from the output of `go help list` command.

**Comment by miiila** [comment]
moving to https://github.com/snyk/snyk-go-plugin/

**Pull request closed without merging**